### PR TITLE
Node 1269

### PR DESCRIPTION
**Pull request by daprahamian**
fix(replset): set server to unknown on not-master
When a not-master occurs, we need to set the server type to
Unknown.

Fixes NODE-1269

No longer WIP, should be good for review.

### CODE REVIEW

File: test/tests/unit/replset/step_down_tests.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
+'use strict';
+
+const expect = require('chai').expect;
+const ReplSet = require('../../../../lib/topologies/replset');
+const mock = require('mongodb-mock-server');
+const ReplSetFixture = require('../common').ReplSetFixture;
+
+describe.only('Step Down (ReplSet)', function() {
```
Comment (jlord) [inline test/tests/unit/replset/step_down_tests.js]: An `only` ✨

Comment (daprahamian) [reply test/tests/unit/replset/step_down_tests.js]: oops

File: lib/topologies/replset.js [hunk 2]
```diff
@@ -1181,37 +1181,59 @@ ReplSet.prototype.getServers = function() {
...
-
-    // Execute the command
-    return self.s.replicaSetState.primary[op](ns, ops, options, callback);
   }
 
-  // increment and assign txnNumber
-  options.txnNumber = getNextTransactionNumber(options.session);
-
-  self.s.replicaSetState.primary[op](ns, ops, options, (err, result) => {
+  const handler = (err, result) => {
```
[the remark this answers is not in the document]
Comment (daprahamian) [reply lib/topologies/replset.js]: It needs to be done in advance b/c of

```js
if (callback.operationId) {
  handler.operationId = callback.operationId;
}
```

File: lib/topologies/replset.js [hunk 3]
```diff
@@ -1227,19 +1249,8 @@ var executeWriteOperation = function(self, op, ns, ops, options, callback) {
...
-  if (this.state === DESTROYED) return callback(new MongoError(f('topology was destroyed')));
-
-  // Not connected but we have a disconnecthandler
-  if (!this.s.replicaSetState.hasPrimary() && this.s.disconnectHandler != null) {
-    return this.s.disconnectHandler.add('insert', ns, ops, options, callback);
-  }
-
   // Execute write operation
-  executeWriteOperation(this, 'insert', ns, ops, options, callback);
+  executeWriteOperation({ self: this, op: 'insert', ns, ops }, options, callback);
```
[the remark this answers is not in the document]
Comment (daprahamian) [reply lib/topologies/replset.js]: I thought of it from a maintenance perspective: once you pass 3 or 4 positional parameters, maintaining code becomes significantly harder.

**Pull request merged**